File: _scheduling_module/scheduling_module/app/core/dcma.py
"""
DCMA 14-Point Schedule Assessment.

Implements the Defense Contract Management Agency 14-point schedule health
checks. Each check returns a pass/fail against its industry threshold plus the
metric value, count, and a corrective observation.

Targets used (standard DCMA):
    1  Logic            < 5% activities missing predecessor OR successor
    2  Leads            0 relationships with negative lag
    3  Lags             < 5% relationships with positive lag
    4  Relationship     >= 90% relationships are Finish-to-Start
    5  Hard Constraints < 5% activities with hard constraints
    6  High Float       < 5% activities with total float > 44 wd
    7  Negative Float   0 activities with total float < 0
    8  High Duration    < 5% incomplete activities with duration > 44 wd
    9  Invalid Dates    0 activities with forecast/actual dates vs data date
    10 Resources        all activities with duration > 0 have cost/resource (info)
    11 Missed Tasks     < 5% activities slipped vs baseline finish
    12 Critical Path    integrity test (does a large delay flow to finish)
    13 CPLI             >= 0.95
    14 BEI              >= 0.95
"""
from __future__ import annotations

from dataclasses import dataclass, field
from datetime import date
from typing import Optional

from .cpm import (CPMActivity, CPMRelationship, CPMEngine, RelType,
                  Constraint, HARD_CONSTRAINTS)
from .calendar import WorkCalendar, DEFAULT_CALENDAR

LONG_DURATION_WD = 44
HIGH_FLOAT_WD = 44


@dataclass
class CheckResult:
    number: int
    name: str
    metric: str          # human-readable value, e.g. "3.2%" or "12"
    threshold: str
    passed: bool
    affected: int
    total: int
    observation: str
    suggestion: str = ""


@dataclass
class DCMAReport:
    checks: list[CheckResult] = field(default_factory=list)
    score: float = 0.0          # % of applicable checks passed
    passed_count: int = 0
    applicable_count: int = 0

    def as_dict(self) -> dict:
        return {
            "score": round(self.score, 1),
            "passed": self.passed_count,
            "applicable": self.applicable_count,
            "checks": [c.__dict__ for c in self.checks],
        }


def _pct(n: int, d: int) -> float:
    return (100.0 * n / d) if d else 0.0


class DCMAAssessor:
    def __init__(
        self,
        activities: list[CPMActivity],
        relationships: list[CPMRelationship],
        project_start: date,
        calendar: WorkCalendar = DEFAULT_CALENDAR,
        data_date: Optional[date] = None,
        baseline_finish: Optional[dict[str, date]] = None,  # activity_id -> BL finish
        baseline_project_finish: Optional[date] = None,
        has_resources: Optional[dict[str, bool]] = None,
    ):
        self.acts = activities
        self.rels = relationships
        self.start = project_start
        self.cal = calendar
        self.data_date = data_date
        self.bl_finish = baseline_finish or {}
        self.bl_proj_finish = baseline_project_finish
        self.has_res = has_resources or {}
        self.by_id = {a.id: a for a in activities}

        # ensure CPM has been run so floats/dates exist
        self.cpm = CPMEngine(activities, relationships, project_start,
                             calendar, data_date).run()

    def assess(self) -> DCMAReport:
        rep = DCMAReport()
        checks = [
            self._c1_logic, self._c2_leads, self._c3_lags, self._c4_relationship,
            self._c5_hard_constraints, self._c6_high_float, self._c7_negative_float,
            self._c8_high_duration, self._c9_invalid_dates, self._c10_resources,
            self._c11_missed_tasks, self._c12_critical_path_test,
            self._c13_cpli, self._c14_bei,
        ]
        for fn in checks:
            r = fn()
            if r:
                rep.checks.append(r)
        applicable = [c for c in rep.checks if c.total >= 0]
        rep.applicable_count = len(applicable)
        rep.passed_count = sum(1 for c in applicable if c.passed)
        rep.score = _pct(rep.passed_count, rep.applicable_count)
        return rep

    # -- individual checks ------------------------------------------------
    def _c1_logic(self) -> CheckResult:
        preds = {r.successor_id for r in self.rels}
        succs = {r.predecessor_id for r in self.rels}
        tasks = [a for a in self.acts]  # incl. milestones
        missing = 0
        for a in tasks:
            no_pred = a.id not in preds
            no_succ = a.id not in succs
            # project start/finish milestones are allowed one open end
            if no_pred or no_succ:
                missing += 1
        p = _pct(missing, len(tasks))
        return CheckResult(1, "Logic", f"{p:.1f}%", "< 5%", p < 5, missing,
                           len(tasks),
                           f"{missing} of {len(tasks)} activities are missing a "
                           f"predecessor and/or successor.",
                           "Add the missing logic links so every activity is "
                           "driven and drives a successor.")

    def _c2_leads(self) -> CheckResult:
        leads = [r for r in self.rels if r.lag < 0]
        return CheckResult(2, "Leads (negative lag)", str(len(leads)), "= 0",
                           len(leads) == 0, len(leads), len(self.rels),
                           f"{len(leads)} relationship(s) use a lead (negative lag).",
                           "Replace leads with explicit logic or break the "
                           "activity into finer steps.")

    def _c3_lags(self) -> CheckResult:
        lags = [r for r in self.rels if r.lag > 0]
        p = _pct(len(lags), len(self.rels))
        return CheckResult(3, "Lags", f"{p:.1f}%", "< 5%", p < 5, len(lags),
                           len(self.rels),
                           f"{len(lags)} of {len(self.rels)} relationships carry "
                           f"a positive lag.",
                           "Model waiting time as real activities instead of lags "
                           "where possible.")

    def _c4_relationship(self) -> CheckResult:
        if not self.rels:
            return CheckResult(4, "Relationship Types", "n/a", ">= 90% FS",
                               True, 0, 0, "No relationships.", "")
        fs = sum(1 for r in self.rels if r.rel_type == RelType.FS)
        p = _pct(fs, len(self.rels))
        return CheckResult(4, "Relationship Types", f"{p:.1f}% FS", ">= 90% FS",
                           p >= 90, len(self.rels) - fs, len(self.rels),
                           f"{p:.1f}% of relationships are Finish-to-Start.",
                           "Prefer FS links; minimise SS/FF/SF which obscure the "
                           "true driving path.")

    def _c5_hard_constraints(self) -> CheckResult:
        hard = [a for a in self.acts if a.constraint_type in HARD_CONSTRAINTS]
        p = _pct(len(hard), len(self.acts))
        return CheckResult(5, "Hard Constraints", f"{p:.1f}%", "< 5%", p < 5,
                           len(hard), len(self.acts),
                           f"{len(hard)} activities use hard constraints "
                           f"(Must Start/Finish On).",
                           "Replace hard constraints with logic; reserve them for "
                           "genuine externally-fixed dates.")

    def _c6_high_float(self) -> CheckResult:
        incomplete = [a for a in self.acts if not a.is_complete]
        hi = [a for a in incomplete
              if a.total_float is not None and a.total_float > HIGH_FLOAT_WD]
        p = _pct(len(hi), len(incomplete) or 1)
        return CheckResult(6, "High Float", f"{p:.1f}%", "< 5%", p < 5, len(hi),
                           len(incomplete),
                           f"{len(hi)} activities have total float > "
                           f"{HIGH_FLOAT_WD} wd.",
                           "High float usually means missing successor logic; "
                           "review the network paths.")

    def _c7_negative_float(self) -> CheckResult:
        neg = [a for a in self.acts
               if a.total_float is not None and a.total_float < 0]
        return CheckResult(7, "Negative Float", str(len(neg)), "= 0",
                           len(neg) == 0, len(neg), len(self.acts),
                           f"{len(neg)} activities have negative total float.",
                           "Recover the schedule or revise constraints/targets "
                           "driving the negative float.")

    def _c8_high_duration(self) -> CheckResult:
        incomplete = [a for a in self.acts
                      if not a.is_complete and not a.is_milestone]
        hi = [a for a in incomplete if a.duration > LONG_DURATION_WD]
        p = _pct(len(hi), len(incomplete) or 1)
        return CheckResult(8, "High Duration", f"{p:.1f}%", "< 5%", p < 5,
                           len(hi), len(incomplete),
                           f"{len(hi)} incomplete activities exceed "
                           f"{LONG_DURATION_WD} wd duration.",
                           "Break long activities into measurable sub-activities "
                           "for better control.")

    def _c9_invalid_dates(self) -> CheckResult:
        if not self.data_date:
            return CheckResult(9, "Invalid Dates", "n/a (no data date)", "= 0",
                               True, 0, 0,
                               "No data date supplied; check skipped.", "")
        bad = 0
        for a in self.acts:
            # forecast (early) start before data date for not-started work
            if not a.is_started and a.es and a.es < self.data_date:
                bad += 1
            # actuals in the future
            if a.actual_start and a.actual_start > self.data_date:
                bad += 1
            if a.actual_finish and a.actual_finish > self.data_date:
                bad += 1
        return CheckResult(9, "Invalid Dates", str(bad), "= 0", bad == 0, bad,
                           len(self.acts),
                           f"{bad} activities have forecast/actual dates "
                           f"inconsistent with the data date.",
                           "Move forecast work to on/after the data date and "
                           "correct any future-dated actuals.")

    def _c10_resources(self) -> CheckResult:
        if not self.has_res:
            return CheckResult(10, "Resources", "n/a", "informational", True, 0, 0,
                               "No resource/cost data provided; check skipped.", "")
        work = [a for a in self.acts if not a.is_milestone and a.duration > 0]
        missing = [a for a in work if not self.has_res.get(a.id, False)]
        p = _pct(len(missing), len(work) or 1)
        return CheckResult(10, "Resources", f"{p:.1f}% unresourced",
                           "all resourced", len(missing) == 0, len(missing),
                           len(work),
                           f"{len(missing)} working activities have no "
                           f"resource/cost assignment.",
                           "Assign resources or cost to enable earned-value "
                           "and BEI tracking.")

    def _c11_missed_tasks(self) -> CheckResult:
        if not self.bl_finish:
            return CheckResult(11, "Missed Tasks", "n/a (no baseline)", "< 5%",
                               True, 0, 0, "No baseline supplied; check skipped.",
                               "")
        considered = [a for a in self.acts if a.id in self.bl_finish]
        missed = 0
        for a in considered:
            bl = self.bl_finish[a.id]
            actual_or_fcast = a.actual_finish or a.ef
            if actual_or_fcast and actual_or_fcast > bl:
                missed += 1
        p = _pct(missed, len(considered) or 1)
        return CheckResult(11, "Missed Tasks", f"{p:.1f}%", "< 5%", p < 5,
                           missed, len(considered),
                           f"{missed} activities finished/forecast later than "
                           f"baseline.",
                           "Investigate slippage causes and update the recovery "
                           "plan.")

    def _c12_critical_path_test(self) -> CheckResult:
        # integrity test: inject a large delay into the first critical activity
        # and confirm the project finish moves by a comparable amount.
        if not self.cpm.critical_path:
            return CheckResult(12, "Critical Path Test", "n/a", "pass", True, 0, 0,
                               "No critical path identified.", "")
        target = self.cpm.critical_path[0]
        bumped = []
        for a in self.acts:
            clone = CPMActivity(
                a.id, a.name, a.duration + (600 if a.id == target else 0),
                a.remaining_duration, a.percent_complete, a.is_milestone,
                a.actual_start, a.actual_finish, a.constraint_type,
                a.constraint_date)
            bumped.append(clone)
        try:
            res2 = CPMEngine(bumped, self.rels, self.start, self.cal,
                             self.data_date).run()
            moved = self.cal.working_days_between(self.cpm.project_finish,
                                                  res2.project_finish) - 1
            ok = moved >= 500   # the 600d bump should flow through
        except Exception:
            ok = False
            moved = 0
        return CheckResult(12, "Critical Path Test", f"+{moved} wd flow",
                           "delay flows to finish", ok, 0, 0,
                           "A large delay injected on the critical path "
                           f"{'correctly' if ok else 'did NOT'} move the project "
                           "finish.",
                           "" if ok else "Broken/soft critical path — check open "
                           "ends and constraints on the driving path.")

    def _c13_cpli(self) -> CheckResult:
        if not self.bl_proj_finish:
            return CheckResult(13, "CPLI", "n/a (no baseline finish)", ">= 0.95",
                               True, 0, 0, "No baseline finish; check skipped.", "")
        cp_len = self.cal.working_days_between(self.start,
                                               self.cpm.project_finish)
        total_float = self.cal.working_days_between(self.cpm.project_finish,
                                                    self.bl_proj_finish)
        # CPLI = (critical path length + project total float) / critical path length
        cpli = (cp_len + total_float) / cp_len if cp_len else 0.0
        return CheckResult(13, "CPLI", f"{cpli:.2f}", ">= 0.95", cpli >= 0.95,
                           0, 0,
                           f"Critical Path Length Index = {cpli:.2f}.",
                           "" if cpli >= 0.95 else "Project is forecast to finish "
                           "later than baseline; assess recovery.")

    def _c14_bei(self) -> CheckResult:
        if not self.bl_finish or not self.data_date:
            return CheckResult(14, "BEI", "n/a", ">= 0.95", True, 0, 0,
                               "Baseline finishes or data date missing; "
                               "check skipped.", "")
        should_complete = [aid for aid, bf in self.bl_finish.items()
                           if bf <= self.data_date]
        completed = sum(1 for aid in should_complete
                        if self.by_id.get(aid) and self.by_id[aid].is_complete)
        denom = len(should_complete) or 1
        bei = completed / denom
        return CheckResult(14, "BEI", f"{bei:.2f}", ">= 0.95", bei >= 0.95,
                           denom - completed, denom,
                           f"Baseline Execution Index = {bei:.2f} "
                           f"({completed}/{denom} due tasks complete).",
                           "" if bei >= 0.95 else "Execution behind baseline plan; "
                           "review productivity and sequencing.")

File: Project-brain/lib/furnace/dcma.ts
"use client";
/**
 * dcma.ts — full DCMA 14-Point Schedule Assessment, clean-room.
 *
 * cpmEngine's dcmaLite covers 10 structural checks. This completes the standard
 * set, including the four that reviewers actually quote in a schedule audit and
 * that dcmaLite could not compute without status/baseline data:
 *
 *   #9  Invalid dates      — actuals in the future, forecasts in the past
 *   #10 Resources          — activities with duration but no resource/cost
 *   #11 Missed tasks       — finished late against baseline, or overdue
 *   #12 Critical path test — inject 600d into a critical activity; if the
 *                            project finish does not move by ~600d the network
 *                            is broken (this is the real integrity test)
 *   #13 CPLI              — Critical Path Length Index, ≥0.95 passes
 *   #14 BEI               — Baseline Execution Index, ≥0.95 passes
 *
 * Thresholds follow the DCMA 14-Point standard as commonly applied. Checks that
 * lack the data to be computed return status "na" rather than a false pass —
 * silently passing an uncomputable check is how schedule audits get gamed.
 */
import { CpmActivity, CpmLink, CpmResult, runCpm } from "@/lib/furnace/cpmEngine";

export type CheckStatus = "pass" | "fail" | "na";

export interface DcmaPoint {
  id: string;
  point: number;
  name: string;
  threshold: string;
  status: CheckStatus;
  count: number;
  total: number;
  pct: number | null;
  value: number | null;      // for index checks (CPLI/BEI)
  offenders: string[];
  detail: string;
}

export interface DcmaReport {
  points: DcmaPoint[];
  passed: number;
  failed: number;
  notApplicable: number;
  score: number;             // % of applicable checks passed
  grade: "A" | "B" | "C" | "D" | "F";
}

/** Status fields an activity may carry; all optional so existing callers compile. */
export interface DcmaActivityExtras {
  actualStart?: string | null;
  actualFinish?: string | null;
  baselineFinish?: string | null;
  resourceCount?: number | null;
  cost?: number | null;
  isMilestone?: boolean;
  isLoE?: boolean;           // level-of-effort activities are exempt from several checks
}

export type DcmaActivity = CpmActivity & DcmaActivityExtras;

const DAY = 86400000;
const HIGH_FLOAT_D = 44;
const LONG_DURATION_D = 44;
const CRITICAL_TEST_INJECT_D = 600;

const pct = (count: number, total: number) => (total > 0 ? (count / total) * 100 : 0);
const codeOf = (a: DcmaActivity) => a.code || a.id;
const linkLabel = (l: CpmLink) => `${l.pred}\u2192${l.succ}`;

function point(
  p: number, id: string, name: string, threshold: string,
  count: number, total: number, limitPct: number | null,
  offenders: string[], detail: string,
): DcmaPoint {
  const percentage = total > 0 ? pct(count, total) : null;
  const status: CheckStatus =
    limitPct === null ? (count === 0 ? "pass" : "fail")
      : percentage === null ? "na"
        : percentage <= limitPct ? "pass" : "fail";
  return {
    id, point: p, name, threshold, status, count, total,
    pct: percentage === null ? null : Math.round(percentage * 10) / 10,
    value: null, offenders: offenders.slice(0, 8), detail,
  };
}

function indexPoint(
  p: number, id: string, name: string, threshold: string,
  value: number | null, min: number, detail: string,
): DcmaPoint {
  return {
    id, point: p, name, threshold,
    status: value === null ? "na" : value >= min ? "pass" : "fail",
    count: 0, total: 0, pct: null,
    value: value === null ? null : Math.round(value * 1000) / 1000,
    offenders: [], detail,
  };
}

/**
 * DCMA #12 — Critical Path Test.
 * Push a large duration into the first critical activity and re-run. A healthy
 * network propagates the whole delay to the project finish; if it absorbs any
 * of it, logic is broken somewhere downstream.
 */
export function criticalPathTest(
  acts: DcmaActivity[], links: CpmLink[], base: CpmResult, dataDate?: string,
): { ok: boolean; expected: number; observed: number; probe: string | null } {
  const probe = acts.find((a) => base.critical.has(a.id) && (a.progress ?? 0) < 100);
  if (!probe) return { ok: false, expected: 0, observed: 0, probe: null };
  const mutated = acts.map((a) =>
    a.id === probe.id ? { ...a, duration: a.duration + CRITICAL_TEST_INJECT_D } : a);
  // dataDate must be carried through or the probe run silently drops every
  // constraint and reports a broken network as healthy.
  const after = runCpm(mutated, links, { dataDate });
  const observed = after.projectDuration - base.projectDuration;
  // allow 1 day of rounding slack
  return {
    ok: Math.abs(observed - CRITICAL_TEST_INJECT_D) <= 1,
    expected: CRITICAL_TEST_INJECT_D,
    observed,
    probe: codeOf(probe),
  };
}

export function runDcma14(
  acts: DcmaActivity[],
  links: CpmLink[],
  result: CpmResult,
  opts: { dataDate?: string; baselineFinish?: string } = {},
): DcmaReport {
  const dataDate = opts.dataDate ? Date.parse(opts.dataDate) : Date.now();
  // LoE and completed milestones distort structural ratios; DCMA excludes them.
  const scope = acts.filter((a) => !a.isLoE);
  const n = scope.length;
  const linkTotal = links.length;

  const hasPred = new Set(links.map((l) => l.succ));
  const hasSucc = new Set(links.map((l) => l.pred));
  const incomplete = scope.filter((a) => (a.progress ?? 0) < 100);

  // ---- 1 Logic -------------------------------------------------------------
  const dangling = scope.filter((a) => !hasPred.has(a.id) || !hasSucc.has(a.id));
  const p1 = point(1, "logic", "Logic — missing predecessor or successor", "\u22645%",
    Math.max(0, dangling.length - 2), n, 5, dangling.map(codeOf),
    "Every activity except one true start and one true finish needs both a predecessor and a successor.");

  // ---- 2 Leads (negative lag) ---------------------------------------------
  const leads = links.filter((l) => l.lag < 0);
  const p2 = point(2, "leads", "Leads — negative lag", "0 allowed",
    leads.length, linkTotal, null, leads.map(linkLabel),
    "Negative lag lets work start before its predecessor finishes and hides true float.");

  // ---- 3 Lags --------------------------------------------------------------
  const lags = links.filter((l) => l.lag > 0);
  const p3 = point(3, "lags", "Lags", "\u22645% of links",
    lags.length, linkTotal, 5, lags.map(linkLabel),
    "Lag is often a substitute for a missing activity; excessive lag makes the network unauditable.");

  // ---- 4 Relationship types ------------------------------------------------
  const nonFs = links.filter((l) => l.type !== "FS");
  const p4 = point(4, "reltypes", "Relationship types — non-FS", "\u226510% must be FS",
    nonFs.length, linkTotal, 10, nonFs.map(linkLabel),
    "At least 90% of links should be Finish-to-Start; SS/FF pairs often mask undefined logic.");

  // ---- 5 Hard constraints --------------------------------------------------
  const hard = scope.filter((a) => a.constraint === "MSO" || a.constraint === "FNLT");
  const p5 = point(5, "constraints", "Hard constraints", "\u22645%",
    hard.length, n, 5, hard.map(codeOf),
    "Mandatory constraints override logic and prevent the network from reacting to delay.");

  // ---- 6 High float --------------------------------------------------------
  const highFloat = incomplete.filter((a) => (result.tf[a.id] ?? 0) > HIGH_FLOAT_D);
  const p6 = point(6, "highfloat", `High float (>${HIGH_FLOAT_D}d)`, "\u22645%",
    highFloat.length, incomplete.length, 5, highFloat.map(codeOf),
    "Float above two months usually indicates missing successor logic rather than genuine slack.");

  // ---- 7 Negative float ----------------------------------------------------
  const negFloat = incomplete.filter((a) => (result.tf[a.id] ?? 0) < 0);
  const p7 = point(7, "negfloat", "Negative float", "0 allowed",
    negFloat.length, incomplete.length, null, negFloat.map(codeOf),
    "Negative float means the schedule is already late against a constraint and needs re-planning.");

  // ---- 8 High duration -----------------------------------------------------
  const longDur = incomplete.filter((a) => a.duration > LONG_DURATION_D && !a.isMilestone);
  const p8 = point(8, "highduration", `High duration (>${LONG_DURATION_D}d)`, "\u22645%",
    longDur.length, incomplete.length, 5, longDur.map(codeOf),
    "Activities longer than one reporting quarter cannot be progressed meaningfully; break them down.");

  // ---- 9 Invalid dates -----------------------------------------------------
  const invalid = scope.filter((a) => {
    const as = a.actualStart ? Date.parse(a.actualStart) : null;
    const af = a.actualFinish ? Date.parse(a.actualFinish) : null;
    // actuals must not be in the future
    if (as !== null && as > dataDate) return true;
    if (af !== null && af > dataDate) return true;
    // an incomplete activity must not claim an actual finish
    if (af !== null && (a.progress ?? 0) < 100) return true;
    // a complete activity must have an actual finish
    if ((a.progress ?? 0) >= 100 && a.actualFinish === null) return true;
    return false;
  });
  const hasStatusData = scope.some((a) => a.actualStart || a.actualFinish);
  const p9 = hasStatusData
    ? point(9, "invaliddates", "Invalid dates", "0 allowed",
      invalid.length, n, null, invalid.map(codeOf),
      "Actual dates beyond the data date, or completed work without an actual finish, invalidate status.")
    : { ...point(9, "invaliddates", "Invalid dates", "0 allowed", 0, 0, null, [], "No actual-date data on this schedule."), status: "na" as CheckStatus };

  // ---- 10 Resources --------------------------------------------------------
  const hasResourceData = scope.some((a) => a.resourceCount != null || a.cost != null);
  const unresourced = scope.filter((a) =>
    !a.isMilestone && a.duration > 0 && !(a.resourceCount || 0) && !(a.cost || 0));
  const p10 = hasResourceData
    ? point(10, "resources", "Resources — activities without resource or cost", "0 allowed",
      unresourced.length, n, null, unresourced.map(codeOf),
      "Unresourced work cannot be levelled or earned against; every non-milestone task needs a resource or cost.")
    : { ...point(10, "resources", "Resources", "0 allowed", 0, 0, null, [], "Schedule is not resource-loaded."), status: "na" as CheckStatus };

  // ---- 11 Missed tasks -----------------------------------------------------
  const withBaseline = scope.filter((a) => a.baselineFinish);
  const missed = withBaseline.filter((a) => {
    const bl = Date.parse(a.baselineFinish!);
    if ((a.progress ?? 0) >= 100 && a.actualFinish) return Date.parse(a.actualFinish) > bl;
    return bl < dataDate; // should have finished by now and has not
  });
  const p11 = withBaseline.length
    ? point(11, "missedtasks", "Missed tasks vs baseline", "\u22645%",
      missed.length, withBaseline.length, 5, missed.map(codeOf),
      "Tasks that finished after baseline, or that should have finished by the data date and have not.")
    : { ...point(11, "missedtasks", "Missed tasks vs baseline", "\u22645%", 0, 0, 5, [], "No baseline finish dates to compare against."), status: "na" as CheckStatus };

  // ---- 12 Critical path test ----------------------------------------------
  const cpt = criticalPathTest(scope, links, result, opts.dataDate);
  const p12: DcmaPoint = {
    id: "criticalpath", point: 12, name: "Critical path test", threshold: "delay must propagate",
    status: cpt.probe === null ? "na" : cpt.ok ? "pass" : "fail",
    count: cpt.ok ? 0 : 1, total: 1, pct: null, value: cpt.observed,
    offenders: cpt.probe ? [cpt.probe] : [],
    detail: cpt.probe === null
      ? "No incomplete critical activity available to probe."
      : `Injected ${cpt.expected}d into ${cpt.probe}; project finish moved ${cpt.observed}d. ` +
        (cpt.ok ? "Network propagates delay correctly."
          : "Delay was absorbed \u2014 broken logic or a constraint is blocking propagation."),
  };

  // ---- 13 CPLI -------------------------------------------------------------
  // CPLI = (critical path length + total float to the finish milestone) / critical path length
  const cpl = result.projectDuration;
  const finishFloat = Math.min(
    0, ...scope.filter((a) => !hasSucc.has(a.id)).map((a) => result.tf[a.id] ?? 0));
  const cpli = cpl > 0 ? (cpl + finishFloat) / cpl : null;
  const p13 = indexPoint(13, "cpli", "CPLI — Critical Path Length Index", "\u22650.95",
    cpli, 0.95,
    cpli === null ? "No critical path length available."
      : `Critical path ${cpl}d with ${finishFloat}d float to completion. Below 0.95 means the finish date is not credible.`);

  // ---- 14 BEI --------------------------------------------------------------
  const baselineDue = withBaseline.filter((a) => Date.parse(a.baselineFinish!) <= dataDate);
  const actuallyDone = scope.filter((a) => a.actualFinish && Date.parse(a.actualFinish) <= dataDate);
  const bei = baselineDue.length ? actuallyDone.length / baselineDue.length : null;
  const p14 = indexPoint(14, "bei", "BEI — Baseline Execution Index", "\u22650.95",
    bei, 0.95,
    bei === null ? "No baseline finish dates due by the data date."
      : `${actuallyDone.length} of ${baselineDue.length} baseline-due activities are complete. Below 0.95 means work is being completed slower than baselined.`);

  const points = [p1, p2, p3, p4, p5, p6, p7, p8, p9, p10, p11, p12, p13, p14];
  const passed = points.filter((p) => p.status === "pass").length;
  const failed = points.filter((p) => p.status === "fail").length;
  const notApplicable = points.filter((p) => p.status === "na").length;
  const applicable = passed + failed;
  const score = applicable ? Math.round((passed / applicable) * 100) : 0;
  const grade: DcmaReport["grade"] =
    score >= 90 ? "A" : score >= 80 ? "B" : score >= 70 ? "C" : score >= 60 ? "D" : "F";

  return { points, passed, failed, notApplicable, score, grade };
}

/** CSV export for the audit pack the Ministry asks for. */
export function dcmaToCsv(report: DcmaReport): string {
  const rows = [
    ["Point", "Check", "Threshold", "Status", "Count", "Total", "Percent", "Value", "Offenders", "Detail"],
    ...report.points.map((p) => [
      String(p.point), p.name, p.threshold, p.status.toUpperCase(),
      String(p.count), String(p.total),
      p.pct === null ? "" : `${p.pct}%`,
      p.value === null ? "" : String(p.value),
      p.offenders.join(" "), p.detail,
    ]),
  ];
  return rows
    .map((r) => r.map((c) => (/[",\n]/.test(c) ? `"${c.replace(/"/g, '""')}"` : c)).join(","))
    .join("\n");
}
